_posts/pl/2021-09-31-zarzadzanie-stanem-aplikacji-frontendowej.md

```diff
@@ -0,0 +1,225 @@
+
+Taki podział store zapewnia bardzo łatwą jego rozszerzalność, gdyż w przypadku dodania nowej funkcjonalności wystarczy dopisać nowy wycinek danych do store. W bardzo dużych aplikacjach, gdzie obiekt stanu aplikacji jest już dość potężny, to podejście ma jeszcze jedną zaletę - pozwala wykorzystać _lazy loading_, to znaczy wczytywać tylko konkretne wycinki store, potrzebne do aktualnie przetwarzanej części aplikacji. 
+
+### Akcje
+Potrzebujemy teraz możliwości wywoływania akcji, żeby powodować zmiany w store. Aplikacja będzie przewidywała dwa zdarzenia: wystosowanie żądania do API po pomysł na jakieś zajęcie oraz dodanie aktywności do listy. Wygląda to następująco:
+
+```typescript
+export const addActivityType = 'Add activity';
+export const getActivityType = 'Get activity';
+
+export const addActivity = createAction(addActivityType, props<ActivityItemModel>());
+export const getActivity = createAction(getActivityType);
+```
+
+Każda akcja musi mieć zdefiniowany swój typ. NgRx używa do tego tzw. _literal type_, czyli po prostu używa typu string do identyfikacji akcji (u nas odpowiednio `Add activity` oraz `Get activity`. W nazywaniu akcji mamy pełną dowolność. Dodatkowo w przypadku dodawania nowej aktywności do listy musimy ją przekazać w ciele akcji przy pomocy metody `props<>()`. Dzięki wykorzystaniu funkcji `createAction` z biblioteki NgRx tworzenie akcji, jak widać, jest bardzo proste.
+
+### Efekt
+Jak wspomniałem, będziemy korzystać z zewnętrznego API. W tym celu napiszemy efekt, który będzie reagował na akcję `Get activity`, następnie wystosowywał żądanie HTTP, a po otrzymaniu odpowiedzi wywoływał akcję `Add Activity`.
+
+```typescript
+@Injectable()
+export class ActivityEffect {
+    constructor(private actions$: Actions,
+                private http: HttpClient) {
+    }
+
+  getActivity$ = createEffect(() => this.actions$.pipe(
+    ofType(getActivityType),
+    mergeMap(() => this.http.get<ActivityItemModel>('https://www.boredapi.com/api/activity')
+      .pipe(
+        map(response => (addActivity(response))),
+        catchError(() => EMPTY)
+      ))
+  ));
+}
+```
+
+Pojawia się tutaj dość dużo zagadnień, zatem spójrzmy na ten fragment kodu od góry do dołu. Najpierw informujemy Angulara, że efekt to tak naprawdę wstrzykiwalna klasa, podobnie jak zwykłe serwisy. W konstruktorze zapewniamy dostęp do serwisu HTTP oraz do zmiennej typu `Actions`. Jest to w gruncie rzeczy serwis udostępniający wszystkie wywołane akcje od momentu ostatniego uruchomienia reducera (czyli dzięki temu damy radę reagować na przychodzące akcje). 
```

Comment:
Nie tyle podobnie co to jest zwykły serwis. Mogę go wstrzyknąć i wołać na nim metody, czy się mylę?

Reply:
Tak. Doprecyzuję